### PR TITLE
Work on verification of votes and ballots.

### PR DESCRIPTION
**Pull request by eb-concordium**
## Purpose

Closes https://github.com/Concordium/concordium-governance-committee-voting/issues/10

## Changes

* Added missing pieces to verification of range proofs.
* Add verification of encrypted ballot.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/eg/src/ballot.rs [hunk 1]
```diff
@@ -139,3 +147,100 @@ impl BallotEncrypted {
             .context("Error writing serialized voter selection to file")
     }
 }
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+pub mod test {
```
Comment (abizjak) [inline src/eg/src/ballot.rs:153]: Does this need to be `pub`?

File: src/eg/src/contest_encrypted.rs [hunk 2]
```diff
@@ -188,4 +203,22 @@ impl ContestEncrypted {
...
+    pub fn verify(&self, header: &PreVotingData, selection_limit: usize) -> bool {
+        let mut j = 1;
+        for ct in &self.selection {
+            let Ok(idx)= Index::from_one_based_index(j as u32) else {return false};
+            let Some(proof) = self.proof_ballot_correctness.get(idx) else {return false};
+            if !ct.verify_ballot_correctness(header, proof) {
+                return false;
+            }
+            j += 1;
+        }
```
Comment (abizjak) [inline src/eg/src/contest_encrypted.rs:216]:
```suggestion
        for (ct, j) in self.selection.iter().zip(1..) {
            let Ok(idx)= Index::from_one_based_index(j) else {return false};
            let Some(proof) = self.proof_ballot_correctness.get(idx) else {return false};
            if !ct.verify_ballot_correctness(header, proof) {
                return false;
            }
        }
```

File: src/eg/src/vec1.rs [hunk 3]
```diff
@@ -10,7 +10,17 @@ use std::collections::TryReserveError;
 use anyhow::{ensure, Error, Result};
 use serde::{Deserialize, Deserializer, Serialize, Serializer};
 
-use crate::index::Index;
+use crate::{index::Index, zk::ProofRange};
+
+pub trait HasIndexType {
```
Comment (abizjak) [inline src/eg/src/vec1.rs:15]: If we're going to keep this new trait we should document it's purpose.

File: src/eg/src/zk.rs [hunk 4]
```diff
@@ -150,6 +153,15 @@ impl ProofRange {
 
         let mut w = <Vec<BigUint>>::with_capacity(big_l + 1);
         for j in 0..big_l + 1 {
+            let c_j_at_least_0 = self.0[j].c >= BigUint::from(0u8);
+            let c_j_less_than_2_256 = self.0[j].c.bits() <= 256; // c_j < 2^256
+            let v_j_at_least_0 = self.0[j].v >= BigUint::from(0u8);
```
Comment (abizjak) [inline src/eg/src/zk.rs:158]: Does this do anything? Since its Uint isn't it >= 0 by construction.

File: src/eg/src/zk.rs [hunk 5]
```diff
@@ -150,6 +153,15 @@ impl ProofRange {
 
         let mut w = <Vec<BigUint>>::with_capacity(big_l + 1);
         for j in 0..big_l + 1 {
+            let c_j_at_least_0 = self.0[j].c >= BigUint::from(0u8);
```
Comment (abizjak) [inline src/eg/src/zk.rs:156]: Does this do anything? Since its Uint isn't it >= 0 by construction.

File: src/eg/src/ballot.rs (hunk 1 above)
Comment (eb-concordium) [reply src/eg/src/ballot.rs:154]: No. Will remove.

File: src/eg/src/zk.rs (hunk 4 above)
Comment (eb-concordium) [reply src/eg/src/zk.rs:158]: Agree. It is redundant, but put it there anyway. Will remove.

File: src/eg/src/zk.rs (hunk 5 above)
Comment (eb-concordium) [reply src/eg/src/zk.rs:156]: I agree.

**Review by abizjak — APPROVED** [review]

**Pull request merged**